metadata/src/main/java/org/apache/kafka/controller/DelegationTokenControlManager.java

```diff
@@ -0,0 +1,359 @@
+        Mac mac = Mac.getInstance("HmacSHA512");
+        SecretKeySpec secretKey = new SecretKeySpec(toBytes(secretKeyString), mac.getAlgorithm());
+
+        mac.init(secretKey);
+        return mac.doFinal(toBytes(tokenId));
+    }
+
+    private TokenInformation getToken(byte[] hmac) {
+        String base64Pwd = Base64.getEncoder().encodeToString(hmac);
+        return tokenCache.tokenForHmac(base64Pwd);
+    }
+
+    private boolean allowedToRenew(TokenInformation tokenInfo, KafkaPrincipal renewer) {
+        if (tokenInfo.owner().equals(renewer)) {
+            return true;
+        }
+        for (KafkaPrincipal validRenewer : tokenInfo.renewers()) {
+            if (validRenewer.equals(renewer)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public boolean isEnabled() {
+        if (secretKeyString != null) {
+            return true;
+        }
+        return false;
+    }
+
+    /*
+     * Pass in the MetadataVersion so that we can return a response to the caller 
+     * if the current metadataVersion is too low.
+     */
+    public ControllerResult<CreateDelegationTokenResponseData> createDelegationToken(
+        ControllerRequestContext context,
+        CreateDelegationTokenRequestData requestData,
+        MetadataVersion metadataVersion
+    ) {
+        long now = time.milliseconds();
+        long maxLifeTime = tokenDefaultMaxLifetime;
+        if (requestData.maxLifetimeMs() > 0) {
+            maxLifeTime = Math.min(maxLifeTime, requestData.maxLifetimeMs());
+        }
+
+        long maxTimestamp = now + maxLifeTime;
+        long expiryTimestamp = Math.min(maxTimestamp, now + tokenDefaultRenewLifetime);
+
+        String tokenId = Uuid.randomUuid().toString();
+
+        KafkaPrincipal owner = context.principal();
+        if ((requestData.ownerPrincipalName() != null) && 
+            (!requestData.ownerPrincipalName().isEmpty())) {
+
+            owner = new KafkaPrincipal(requestData.ownerPrincipalType(), requestData.ownerPrincipalName());
+        }
+        CreateDelegationTokenResponseData responseData = new CreateDelegationTokenResponseData()
+                .setPrincipalName(owner.getName())
+                .setPrincipalType(owner.getPrincipalType())
+                .setTokenRequesterPrincipalName(context.principal().getName())
+                .setTokenRequesterPrincipalType(context.principal().getPrincipalType());
+
+        List<ApiMessageAndVersion> records = new ArrayList<>();
+
+        if (secretKeyString == null) {
+            // DelegationTokens are not enabled
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(DELEGATION_TOKEN_AUTH_DISABLED.code()));
+        }
+
+        if (!metadataVersion.isDelegationTokenSupported()) {
+            // DelegationTokens are not supported in this metadata version
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(UNSUPPORTED_VERSION.code()));
+        }
```

Comment:
Should go at top of method?

Reply:
The create response needs an owner and requester. 

Should the UNSUPPORTED_VERSION check happen before the AUTH_DISABLED check? I think AUTH_DSABLED check should be first.

Reply:
Agreed, AUTH_DSABLED check should be first